### PR TITLE
Concatenation over epsilon transitions

### PR DESCRIPTION
**Pull request by Adda0**
This PR implements concatenation over epsilon transitions.

Furthermore, `tests-nfa.cc`was becoming too long and it was getting harder to find required tests in the file. I extracted tests for segmentation and concatenation to their own files and extended both tests further.

The PR also includes fixes for `remove_epsilon()`, forces results of concatenation to be empty (without any states at all) for empty concatenations and fixes setting final and initial states for segments in segmentation.

### CODE REVIEW

**Review by martinhruska — APPROVED** [review]
I have no problems with PR. I back spli of nfa.cc to more files, it will be clearer now.

File: include/mata/nfa.hh [hunk 1]
```diff
@@ -1168,6 +1186,11 @@ private:
      */
     void handle_epsilon_transitions(const StateDepthPair& state_depth_pair, const TransSymbolStates& state_transitions,
                                     std::deque<StateDepthPair>& worklist);
+
+    /**
+     * Clear initial states for all segment but the first one. Clear final states for all segments but the last one.
+     */
+    void clear_initial_final_states();
```
Comment (tfiedor) [inline include/mata/nfa.hh:1193]: I find this naming confusing, as it doesn't clear all initial and final states, nor there is something like `initial final` states. I have no idea, what would be better though. Maybe `adjust_initial_and_final_states` or `update_initial_and_final_states`?

File: src/nfa/nfa-concatenation.cc [hunk 2]
```diff
@@ -70,37 +80,72 @@ class Concatenation
...
         add_lhs_transitions();
         make_final_states();
         add_rhs_transitions();
     }
 
-    /**'
+    /**
+     * Compute concatenation of given automata concatenating over epsilon transitions.
+     */
+    void compute_epsilon_concatenation() {
```
Comment (tfiedor) [inline src/nfa/nfa-concatenation.cc:104]: Maybe rename to `compute_concatenation_with_epsilon` or `compute_concatenation_with_epsilon_transitions`?

**Review by tfiedor — APPROVED** [review]
No issues, just two naming suggestion (or rather opportunities for better naming ;) )

File: src/nfa/nfa-concatenation.cc (hunk 2 above)
Comment (Adda0) [reply src/nfa/nfa-concatenation.cc:104]: The best name exactly describing the operation is `compute_concatenation_over_epsilon_transitions`. What do you say?

File: include/mata/nfa.hh (hunk 1 above)
Comment (Adda0) [reply include/mata/nfa.hh:1193]: The function removes *inner* initial and final states. Then, something like `remove_inner_initial_final_states` could be used.

File: src/nfa/nfa-concatenation.cc (hunk 2 above)
Comment (tfiedor) [reply src/nfa/nfa-concatenation.cc:104]: Sounds good.

File: include/mata/nfa.hh (hunk 1 above)
Comment (tfiedor) [reply include/mata/nfa.hh:1193]: Sounds better, and maybe add `and` :)

Comment (Adda0) [reply include/mata/nfa.hh:1193]: Great. Will adjust accordingly in the code.

File: src/nfa/nfa-concatenation.cc (hunk 2 above)
Comment (martinhruska) [reply src/nfa/nfa-concatenation.cc:104]: I think it's too loooooooong. I like the original version more, however this is probably better understandable so even someone after braing surgery would know what function does.

Comment (tfiedor) [reply src/nfa/nfa-concatenation.cc:104]: Or maybe just `compute_concatenation_over_epsilon`?

Comment (martinhruska) [reply src/nfa/nfa-concatenation.cc:104]: or `concatenate_over_epsilon`?

Comment (tfiedor) [reply src/nfa/nfa-concatenation.cc:104]: No issues with thtat

Comment (Adda0) [reply src/nfa/nfa-concatenation.cc]: `concatenate_over_epsilon` can be used, too.

**Comment by Adda0** [comment]
Both names were changed as agreed upon. When all checks finish, the PR is ready to be merged.

**Pull request merged**